core/src/test/scala/unit/kafka/server/KafkaServerTest.scala

```diff
@@ -154,6 +159,25 @@ class KafkaServerTest extends QuorumTestHarness {
     server.shutdown()
   }
 
+  @Test
+  def testOfflineDirectoryOnMetadataCheckpointIOError(): Unit = {
+    val props = TestUtils.createBrokerConfigs(1, zkConnect).head
+    val kafkaConfig = KafkaConfig.fromProps(props)
+    val server: KafkaServer = TestUtils.createServer(kafkaConfig, new MockTime, None, startup = false)
+
+    val spyValue: BrokerMetadataCheckpoint = spy(server.brokerMetadataCheckpoints.head._2)
+    doThrow(new IOException()).when(spyValue).write(any())
+    server.brokerMetadataCheckpoints = server.brokerMetadataCheckpoints.updated(server.brokerMetadataCheckpoints.head._1, spyValue)
+
+    try {
+      server.startup()
+      assertTrue(server.logDirFailureChannel.hasOfflineLogDir(kafkaConfig.logDirs.head),
```

Comment:
Could this flake? Might be worth enclosing in `TestUtils.waitUntilTrue` perhaps?